### PR TITLE
Retry policy will always clone the *http.Request

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
This ensures that the entirety of the request is restored on retries.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/azcore/runtime/policy_retry_test.go [hunk 1]
```diff
@@ -810,6 +810,47 @@ func TestRetryableRequestBodyWithCloser(t *testing.T) {
...
+	if req.Body() == nil {
+		return nil, errors.New("request body wasn't restored")
+	}
+	if req.Raw().Header.Get("content-type") != "text/plain" {
+		return nil, errors.New("content-type header wasn't restored")
+	}
+
+	if body := req.Body(); body != nil {
+		rb := reqBody{body, req.Raw().Header.Get("content-type")}
+		req.SetOperationValue(rb)
```
Comment (chlowell) [inline sdk/azcore/runtime/policy_retry_test.go:846]: My first thought was, the actual policy does this but it isn't necessary for this test. Now I'm wondering whether the OperationValue feature needs additional testing. The actual policy's logic is:
1. set: `req.SetOperationValue()`
2. send: `req.Next()`
3. get: `req.OperationValue()`

It appears every retry runs on its own clone of `req`, and a policy never gets `req` itself. So step 1 sets the value on an ephemeral clone; if there's a retry during step 2, is that value lost?

Comment (chlowell) [reply sdk/azcore/runtime/policy_retry_test.go:846]: I see. Then I'm back to wondering why this test sets the value 😆. Does it have a side effect?

**Review by chlowell — APPROVED** [review]

File: sdk/azcore/runtime/policy_retry_test.go (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/policy_retry_test.go:846]: No, it's just copy/pasta from the key vault challenge policy.  I'll clean this up a bit.

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/policy_retry_test.go]: Done.

**Pull request merged**